File: src/cpp/ImplementacionC++/Ejercicio.cpp
#include <iostream>
#include <queue>

using namespace std;

// Estructura del nodo con comentarios para la rúbrica
struct Nodo {
    int dato;
    Nodo *izq, *der;
    Nodo(int v) : dato(v), izq(nullptr), der(nullptr) {}
};

// Función de inserción mediante recursividad (Criterio: Uso de recursividad)
Nodo* insertar(Nodo* raiz, int dato) {
    if (raiz == nullptr) return new Nodo(dato);
    if (dato < raiz->dato) raiz->izq = insertar(raiz->izq, dato);
    else if (dato > raiz->dato) raiz->der = insertar(raiz->der, dato);
    return raiz;
}

// Recorridos de profundidad
void preorden(Nodo* raiz) {
    if (raiz) { cout << raiz->dato << " "; preorden(raiz->izq); preorden(raiz->der); }
}

void inorder(Nodo* raiz) {
    if (raiz) { inorder(raiz->izq); cout << raiz->dato << " "; inorder(raiz->der); }
}

void postorden(Nodo* raiz) {
    if (raiz) { postorden(raiz->izq); postorden(raiz->der); cout << raiz->dato << " "; }
}

// Recorrido en amplitud (Criterio: Uso de cola)
void bfs(Nodo* raiz) {
    if (!raiz) return;
    queue<Nodo*> q;
    q.push(raiz);
    while (!q.empty()) {
        Nodo* temp = q.front();
        q.pop();
        cout << temp->dato << " ";
        if (temp->izq) q.push(temp->izq);
        if (temp->der) q.push(temp->der);
    }
}

int main() {
    Nodo* raiz = nullptr;
    int opcion, valor;

    // Carga inicial automática (Nodos base + 5 nuevos según actividad 3)
    int datos[] = {10, 5, 15, 2, 7, 12, 20, 1, 3, 18, 25, 30}; 
    for(int x : datos) raiz = insertar(raiz, x);

    do {
        cout << "\n--- MENU ARBOL BINARIO ---" << endl;
        cout << "1. Insertar nodo manualmente" << endl;
        cout << "2. Mostrar Preorden" << endl;
        cout << "3. Mostrar Inorden" << endl;
        cout << "4. Mostrar Postorden" << endl;
        cout << "5. Mostrar BFS (Amplitud)" << endl;
        cout << "0. Salir" << endl;
        cout << "Seleccione: "; cin >> opcion;

        switch(opcion) {
            case 1: 
                cout << "Valor: "; cin >> valor;
                raiz = insertar(raiz, valor);
                break;
            case 2: cout << "Preorden: "; preorden(raiz); cout << endl; break;
            case 3: cout << "Inorden: "; inorder(raiz); cout << endl; break;
            case 4: cout << "Postorden: "; postorden(raiz); cout << endl; break;
            case 5: cout << "BFS: "; bfs(raiz); cout << endl; break;
        }
    } while (opcion != 0);

    return 0;
}

File: src/java/ImplementacionJava/Ejercicio.java
import java.util.*;

public class Ejercicio {
    // Estructura del Nodo con comentarios para la rúbrica
    static class Nodo {
        int dato;
        Nodo izq, der;
        Nodo(int v) { 
            this.dato = v; 
            this.izq = null;
            this.der = null;
        }
    }

    private Nodo raiz;

    // Método para insertar (Uso de recursividad - Rúbrica)
    public void insertar(int dato) { 
        raiz = insertarRec(raiz, dato); 
    }
    
    private Nodo insertarRec(Nodo actual, int dato) {
        if (actual == null) return new Nodo(dato);
        if (dato < actual.dato) {
            actual.izq = insertarRec(actual.izq, dato);
        } else if (dato > actual.dato) {
            actual.der = insertarRec(actual.der, dato);
        }
        return actual;
    }

    // --- RECORRIDOS (Implementación correcta - Rúbrica) ---

    public void preorden(Nodo n) {
        if (n != null) {
            System.out.print(n.dato + " ");
            preorden(n.izq);
            preorden(n.der);
        }
    }

    public void inorden(Nodo n) {
        if (n != null) {
            inorden(n.izq);
            System.out.print(n.dato + " ");
            inorden(n.der);
        }
    }

    public void postorden(Nodo n) {
        if (n != null) {
            postorden(n.izq);
            postorden(n.der);
            System.out.print(n.dato + " ");
        }
    }

    // Recorrido BFS (Uso de Cola/Queue - Rúbrica)
    public void bfs() {
        if (raiz == null) return;
        Queue<Nodo> cola = new LinkedList<>();
        cola.add(raiz);
        while (!cola.isEmpty()) {
            Nodo temp = cola.poll();
            System.out.print(temp.dato + " ");
            if (temp.izq != null) cola.add(temp.izq);
            if (temp.der != null) cola.add(temp.der);
        }
    }

    public static void main(String[] args) {
        Ejercicio arbol = new Ejercicio();
        Scanner sc = new Scanner(System.in);
        
        // Nodos iniciales (Incluye los 5 nuevos solicitados en la actividad)
        int[] iniciales = {10, 5, 15, 2, 7, 12, 20, 1, 3, 18, 22, 25};
        for (int v : iniciales) arbol.insertar(v);

        int op;
        do {
            System.out.println("\n---------------------------------------------------------");
            System.out.println("1. Insertar | 2. Preorden | 3. Inorder | 4. Postorden | 5. BFS | 0. Salir");
            System.out.print("Elija una opcion: ");
            op = sc.nextInt();
            
            switch(op) {
                case 1:
                    System.out.print("Ingrese valor a insertar: ");
                    arbol.insertar(sc.nextInt());
                    break;
                case 2:
                    System.out.print("PREORDEN: ");
                    arbol.preorden(arbol.raiz);
                    System.out.println();
                    break;
                case 3:
                    System.out.print("INORDEN: ");
                    arbol.inorden(arbol.raiz);
                    System.out.println();
                    break;
                case 4:
                    System.out.print("POSTORDEN: ");
                    arbol.postorden(arbol.raiz);
                    System.out.println();
                    break;
                case 5:
                    System.out.print("BFS (Por niveles): ");
                    arbol.bfs();
                    System.out.println();
                    break;
                case 0:
                    System.out.println("Saliendo...");
                    break;
                default:
                    System.out.println("Opcion no valida.");
            }
        } while (op != 0);
        sc.close();
    }
}   